[DigitalOcean] droplet integration

asaiacai opened the pull request:
<!-- Describe the changes in this PR -->

This adds digital ocean droplets to the sky.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] All smoke tests: `pytest tests/test_smoke.py --do`

moinnadeem commented:
@asaiacai I would be really interested in using this work -- do you know when you might be able to land this PR?

asaiacai commented:
Hello! I’m aiming for this to land in next two weeks providing testing goes
smoothly. Stay tuned!

2024年8月30日(金) 22:22 Moin Nadeem ***@***.***>:

> @asaiacai <https://github.com/asaiacai> I would be really interested in
> using this work -- do you know when you might be able to land this PR?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skypilot-org/skypilot/pull/3832#issuecomment-2322778392>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGAWAHWFM5NGH6FGYPGAC4LZUFHIVAVCNFSM6AAAAABMRCIIO6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMRSG43TQMZZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

asaiacai commented:
@Michaelvll i think this PR is finally ready, once you have a free moment! I've currently disabled tests for gpu droplets as they are still in early access, but I think I got the remaining tests to work on normal CPU droplets fine. Let me know if you find any issues.

asaiacai commented:
also this catalog update is required to pass the `sky serve` requests since I added cheaper `2vcpu` instances.

skypilot-org/skypilot-catalog#83

cblmemo commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,275 @@
+    """Digital Ocean Cloud"""
+
+    _REPR = 'DO'
+    _CLOUD_UNSUPPORTED_FEATURES = {
```
Do we supprot opening ports on DO? Seems like we don't have corresponding implementation in provisioner.

asaiacai replied:
ports are exposed by default on digital ocean instances so you don't need any further action to open ports. resolving

cblmemo replied:
It will be good to add a comment here ;)

https://github.com/skypilot-org/skypilot/blob/800f7d6971bd604f266faebb33d044c7d5baca55/sky/provision/fluidstack/instance.py#L331-L332

cblmemo reviewed (commented):
Thanks for this amazing work @asaiacai ! This would be very helpful. Left some discussions ;)

cblmemo commented on docs/source/getting-started/installation.rst:
Actually, after a second glance, it seems like this is the official abbreviation used in their python package. Lets keep the original name, but maybe we should make the cloud argument to `digitalocean`, i.e. `sky launch --cloud digitalocean`?

asaiacai replied:
added `digitalocean` as alias

asaiacai commented on sky/adaptors/do.py:
```diff
@@ -0,0 +1,18 @@
+_LAZY_MODULES = (pydo, azure)
+
+
+@common.load_lazy_modules(modules=_LAZY_MODULES)
+def exceptions():
+    """Azure exceptions."""
+    from azure.core import exceptions as azure_exceptions
+    return azure_exceptions
```
hahaha you're reaction is the same as mine. The reason is `pydo` (for some reason) uses the `azure.core` library and inherits the same exception classes. It's even in one of their examples. 
```
# Would be nice to not need azure branded imports.
from azure.core.exceptions import HttpResponseError
```

https://github.com/digitalocean/pydo/blob/7b01498d99eb0d3a772366b642e5fab3d6fc6aa2/examples/poc_droplets_volumes_sshkeys.py#L6

cblmemo replied:
wow interesting. A comment here would help reduce confusion ;)

asaiacai commented on sky/templates/do-ray.yml.j2:
```diff
@@ -0,0 +1,92 @@
+    sudo bash -c 'rm -rf /etc/security/limits.d; echo "* soft nofile 1048576" >> /etc/security/limits.conf; echo "* hard nofile 1048576" >> /etc/security/limits.conf';
+    sudo grep -e '^DefaultTasksMax' /etc/systemd/system.conf || (sudo bash -c 'echo "DefaultTasksMax=infinity" >> /etc/systemd/system.conf'); sudo systemctl set-property user-$(id -u $(whoami)).slice TasksMax=infinity; sudo systemctl daemon-reload;
+    mkdir -p ~/.ssh; (grep -Pzo -q "Host \*\n  StrictHostKeyChecking no" ~/.ssh/config) || printf "Host *\n  StrictHostKeyChecking no\n" >> ~/.ssh/config;
+    [ -f /etc/fuse.conf ] && sudo sed -i 's/#user_allow_other/user_allow_other/g' /etc/fuse.conf || (sudo sh -c 'echo "user_allow_other" > /etc/fuse.conf');
```
sounds good. done.

asaiacai commented on tests/test_smoke.py:
```diff
@@ -1341,6 +1342,7 @@ def test_scp_logs():
 @pytest.mark.no_scp  # SCP does not have T4 gpus. Run test_scp_job_queue instead
 @pytest.mark.no_paperspace  # Paperspace does not have T4 gpus.
 @pytest.mark.no_oci  # OCI does not have T4 gpus
+@pytest.mark.no_do  # DO does not have T4 gpus
```
adding a test case for h100 (the only supported gpu type)

asaiacai commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,275 @@
+            'Docker container images as runtime environments'
+            f' are not supported in {_REPR}. Try using in `run`',
+    }
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 120
```
added comment linking to docs
https://docs.digitalocean.com/reference/api/api-reference/#operation/droplets_create

cblmemo replied:
We will also append some suffix (e.g. '-worker`) to the cluster name so we would need a calculation for the actual cluster name limit. Reference here: https://github.com/skypilot-org/skypilot/blob/e870839aeed16c118c0eb1f4889efc20006c27c4/sky/clouds/aws.py#L103-L108

cblmemo commented on sky/provision/do/utils.py:
```diff
@@ -0,0 +1,209 @@
+        'name': instance_name,
+        'region': region,
+        'filesystem_type': 'ext4',
+        'tags': ['skypilot', cluster_name_on_cloud]
```
`key:value` sounds like a good idea! It will be great to include tags in this PR as well :))

asaiacai commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,275 @@
+        clouds.CloudImplementationFeatures.CUSTOM_DISK_TIER:
+            'Custom disk tiers'
+            f' is not supported in {_REPR}.',
+        clouds.CloudImplementationFeatures.DOCKER_IMAGE:
+            'Docker container images as runtime environments'
+            f' are not supported in {_REPR}. Try using in `run`',
```
I included a startup script that installs docker, but ssh comes up before the startup script finishes running resulting in the docker environment setup failing. Docker does eventually get setup on the machine. If I for example, reattempt to launch on the same cluster after waiting for like 10 seconds, I can successfully use the docker environment, but the times i tested it never worked on freshly provisioned instances.

cblmemo replied:
Is it possible to use a base image with docker installed? Or, can we wait for the docker to finish setup for DO cloud? 

https://github.com/skypilot-org/skypilot/blob/e870839aeed16c118c0eb1f4889efc20006c27c4/sky/provision/docker_utils.py#L336-L343

asaiacai replied:
i added waiting for `docker` to `_check_docker_installed` but let me know if this should only be specific to `DO`

cblmemo replied:
My concern is that it will take too long to error out for an image without docker installation. Is it possible that we somehow detect if the docker installation is in progress, and immediately return if we found no installation?

asaiacai replied:
@cblmemo the most recent AI/ML images from digital ocean come with docker and are available now for CPU vms as well gpu vms. i'm going to revert the docker wait and just set this as the default image.

asaiacai commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,275 @@
+                                            accelerators,
+                                            use_spot,
+                                            region=region,
+                                            zone=None)
```
in their API spec/docs they  refer to for example `nyc1`, `nyc2` and `nyc3` as distinct regions but don't use the term zone, but perhaps it makes more sense to refer to these as zones in the context of skypilot. lmk what you think

cblmemo replied:
Lets just keep it as regions!

asaiacai commented on tests/skyserve/update/bump_version_after.yaml:
resolved by #3956

asaiacai commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,275 @@
+                cpus=resources.cpus,
+                memory=resources.memory,
+                disk_tier=resources.disk_tier)
+            if default_instance_type is None:
+                return resources_utils.FeasibleResources([], [], None)
```
i copied this over from the paperspace provisioner, but this code path should never be it and isn't when I test `sky launch`. I'm going to remove it.

asaiacai commented on sky/provision/do/instance.py:
```diff
@@ -0,0 +1,299 @@
+        if worker_only and instance_name.endswith('-head'):
+            continue
+        try:
+            utils.client().droplets.destroy_with_associated_resources_dangerous(
+                droplet_id=instance_meta['id'], x_dangerous=True)
```
+1 to wrapping the low level APIs. we attach block storage to each instance so that `disk_size` is actually usable, otherwise each instance only has a fixed disk size that's tied to the instance type. we use `destroy_with_associated_resources_dangerous` since this atomically cleans up the block storage and the instance, without which, we can end up with resource leakage of the associated block storage in an autodown context.

cblmemo replied:
Good point! It will be good to have a comment for it for future reference.

asaiacai commented on sky/provision/do/instance.py:
```diff
@@ -0,0 +1,299 @@
+        instances[instance_name] = [
+            common.InstanceInfo(
+                instance_id=instance_meta['name'],
+                internal_ip=instance_ip,
+                external_ip=instance_ip,
```
the short answer is multinode breaks if we use the internal ip here (I already tried this).

the long answer is the public ip address is exposed as the first ip address on every node. The ray executor compares the `internal_ip` against the one's ray discovers. However, ray will discover the first ip address it sees on each node, which is `eth0` ip address which is the public one, the executor will see that there is a mismatch in the internal ips given by skypilot and the ones ray discovers and execution will fail.

```
(base) root@sky-eb1b-gcpuser-323b-0a28-head:~# ip -br addr
lo               UNKNOWN        127.0.0.1/8 ::1/128 
eth0             UP             167.172.255.36/20 10.17.0.5/16 fe80::a017:c0ff:fe07:6e1f/64 
eth1             UP             10.108.0.2/20 fe80::a0db:f7ff:fea5:82e3/64 
docker0          DOWN           172.17.0.1/16
```

cblmemo replied:
So ray will always treat eth0 as the private ip address? is there any method to customize it?

asaiacai replied:
i want to say no. we use `ray.util.get_node_ip_address` [here](https://github.com/skypilot-org/skypilot/blob/057bc4b44755ac1e9dadc680e022c369e8ddff52/sky/backends/cloud_vm_ray_backend.py#L499), and it doesn't really expose any arguments that would let us bypass the first detected ip interface. 
https://github.com/ray-project/ray/blob/f44890f90284c41c4d26a2d1e3b36b3ce4a8f4ff/python/ray/_private/services.py#L677

I also tried taking down the `eth0` interface, but this will make the instance unreachable from the public internet

cblmemo replied:
I see. looks good to me. cc @Michaelvll for a look here

cblmemo commented on sky/provision/do/utils.py:
```diff
@@ -0,0 +1,209 @@
+    instance = _create_droplet(instance_request)
+
+    volume_request = {
+        'size_gigabytes': config.node_config['DiskSize'],
```
quick question: Does DO support multiple disk tier as well?

asaiacai replied:
no there's only one disk tier.

asaiacai commented on sky/provision/do/utils.py:
```diff
@@ -0,0 +1,209 @@
+INITIAL_BACKOFF_SECONDS = 10
+MAX_BACKOFF_FACTOR = 10
+MAX_ATTEMPTS = 6
+SSH_KEY_NAME = f'sky-key-{common_utils.get_user_hash()}'
```
`SSH_KEY_NAME` refers to the name of the key as stored by the digital ocean API, not the actual name of the key file stored on machine doing `sky launch`. The actual keys are passed in via
```
ssh_key_id(config.authentication_config['ssh_public_key'])['fingerprint']
```
during instance creation with `config.authentication_config['ssh_public_key']` coming from `get_or_generate_keys` already.

`SSH_KEY_NAME = f'sky-key-{common_utils.get_user_hash()}'` exists so that if multiple people share the same account but are launching across different machines, there isn't a collision in the API around the SSH key. lmk what you think

cblmemo replied:
Sounds good! If that is the case, we could consider rename it to sth like `SSH_KEY_NAME_ON_DO` to reduce confusion ; )

asaiacai commented on sky/provision/do/instance.py:
```diff
@@ -0,0 +1,299 @@
+        utils.client().droplet_actions.post(droplet_id=instance_meta['id'],
+                                            body={'type': 'power_on'})
+    for _ in range(MAX_POLLS_FOR_UP_OR_STOP):
+        instances = utils.filter_instances(cluster_name_on_cloud,
+                                           pending_status + ['off'])
```
ah good catch. `pending_status` is only `['new']` which is incorrect to include for restarting stopped instnaces. removing.

asaiacai commented on sky/provision/do/instance.py:
```diff
@@ -0,0 +1,299 @@
+        time.sleep(constants.POLL_INTERVAL)
+    else:
+        # Failed to launch config.count of instances after max retries
+        msg = ('run_instances: Failed to create the'
+               'instances due to capacity issue.')
+        logger.warning(msg)
```
Reading this again, I realize capacity failures will be handled  in `utils` by the exception handler for `_create_droplet`. However, we should keep a failure message to indicate that something had failed with provisioning, but it will be hard to catch if it occurs somehow outside of the `_create_droplet` invocation.

asaiacai commented:
@cblmemo reran smoke, but if you have a moment to review again, thanks in advance!

asaiacai commented:
@cblmemo digital ocean released a new image for AI/ML with docker. It's a lot more bloated than the `ubuntu-22.04` images that were previously used for the CPU VMs but it does ship with docker, and i've removed the docker wait. please let me know if you need anything else for this PR.

cblmemo commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,297 @@
+            else:
+                assert region.name in resources.image_id
+                image_id = resources.image_id[region.name]
+        return {
+            'instance_type': resources.instance_type,
+            'custom_resources': custom_resources,
+            'region': region.name,
+            **({
+                'image_id': image_id
+            } if image_id else {})
+        }
```
```suggestion
        resources_vars = {
            'instance_type': resources.instance_type,
            'custom_resources': custom_resources,
            'region': region.name,
        }
        if image_id is not None:
            resources_vars['image_id'] = image_id
        return resources_vars
```

cblmemo commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,297 @@
+        return True, None
+
+    def get_credential_file_mounts(self) -> Dict[str, str]:
+        do_utils.client()
```
```suggestion
```
accidentally?

asaiacai replied:
no you need this to initialize `do_utils.CREDENTIALS_PATH`. I've added a comment explaining why

cblmemo commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,297 @@
+        # Can implement later if desired
+        return None
+
+    @classmethod
+    def get_image_size(cls, image_id: str, region: Optional[str]) -> float:
+        del region
+        try:
+            response = do_utils.client().images.get(image_id=image_id)
+            return response['image']['size_gigabytes']
+        except do.exceptions().HttpResponseError as err:
+            raise do_utils.DigitalOceanError(
+                'HTTP error while retrieving size of '
+                f'image_id {response}: {err.error.message}') from err
+        except KeyError as err:
+            raise do_utils.DigitalOceanError(
+                f'No image_id `{image_id}` found') from err
```
```suggestion
    @classmethod
    def get_image_size(cls, image_id: str, region: Optional[str]) -> float:
        del region
        try:
            response = do_utils.client().images.get(image_id=image_id)
            if not response:
                raise do_utils.DigitalOceanError(
                    f'No image_id `{image_id}` found') from err
            return response['image']['size_gigabytes']
        except do.exceptions().HttpResponseError as err:
            raise do_utils.DigitalOceanError(
                'HTTP error while retrieving size of '
                f'image_id {response}: {err.error.message}') from err
```

cblmemo commented on sky/clouds/service_catalog/do_catalog.py:
Should we add a todo here to add fetchers for DO?

cblmemo commented on sky/templates/do-ray.yml.j2:
```diff
@@ -0,0 +1,98 @@
+{%- endfor %}
+}
+
+rsync_exclude: []
+
+initialization_commands: []
+
```
```suggestion
```
seems like we dont need this two?

cblmemo commented on tests/test_smoke.py:
```diff
@@ -1559,6 +1588,50 @@ def test_job_queue_with_docker(generic_cloud: str, image_id: str):
     run_one_test(test)
 
 
+@pytest.mark.do
+def test_do_job_queue_with_docker(generic_cloud: str, image_id: str):
```
Is it possible to reuse the previous function and add another argument for accelerator to use? same for other tests in this file. You can checkout `@pytest.mark.parametrize`

cblmemo reviewed (commented):
Thanks @asaiacai for this amazing contribution! Looks mostly good to me. After those nits get resolved and smoke test passed, it should be ready to go!

asaiacai commented on sky/clouds/do.py:
```diff
@@ -0,0 +1,297 @@
+        return 0.0
+
+    def get_egress_cost(self, num_gigabytes: float) -> float:
+        return 0.0
```
first 500GB/month free afterwards its $0.01/GB excess. added a comment

cblmemo commented:
> Hi @asaiacai , I'm considering adding a new cloud provider to Sky and was reviewing the Google Doc linked in Sky's documentation. However, I couldn't find any guidance on setting up test files for this. I looked through the [test folder](https://github.com/skypilot-org/skypilot/tree/master/tests) and the [GitHub Actions file](https://github.com/skypilot-org/skypilot/blob/master/.github/workflows/pytest.yml), but I'm unclear about how cloud credentials are configured for tests in GitHub CI. Could you help me understand how Sky sets up the tests for CI? I didn't see any references to GitHub secrets in the actions file, so I feel like I'm missing some information.
> 
> cc: @cblmemo

Hi @strict-type , the tests in github actions does not requires cloud credentials. All tests require credentials is located in [test_smoke.py](https://github.com/skypilot-org/skypilot/blob/master/tests/test_smoke.py), which needs to be ran locally and leverage your local credentials.

asaiacai commented:
@cblmemo if you have a spare cycle

cblmemo commented:
> @cblmemo if you have a spare cycle

Hi @asaiacai, It seems like the CI is failing. Could you help take a look at what is happening here?

cblmemo commented on sky/clouds/service_catalog/do_catalog.py:
```diff
@@ -0,0 +1,113 @@
+query instance types and pricing information for digital ocean.
+"""
+
+#TODO(asaiacai): add fetchers
```
```suggestion
# TODO(asaiacai): add fetchers
```

cblmemo commented on sky/provision/do/utils.py:
```diff
@@ -0,0 +1,309 @@
+            logger.debug(f'Digital Ocean credential path found at {path}')
+    if not credentials_found == 1:
+        raise DigitalOceanError(
+            (f'{credentials_found} credentials '
+             'found, expected == 1, ',
+             'try setting a new API token with `doctl auth init`'))
```
```suggestion
            (f'{credentials_found} credentials '
             'found. Expected 1 credential. Please ',
             'try setting a new API token with `doctl auth init`'))
```

cblmemo reviewed (commented):
Thanks @asaiacai ! Mostly looks good to me. Left 2 nits. If the CI and smoke test passed, it should be ready to go!

cblmemo commented on sky/provision/do/utils.py:
```diff
@@ -90,9 +90,8 @@ def _init_client():
-            raise DigitalOceanError(
-                'no valid api tokens found try '
-                'setting a new API token with `doctl auth init`')
+            logger.debug('no valid api tokens found try '
+                         'setting a new API token with `doctl auth init`')
```
should

cg505 commented on sky/setup_files/setup.py:
```diff
@@ -139,128 +139,6 @@ def parse_readme(readme: str) -> str:
-    'runpod': ['runpod>=1.5.1'],
-    'fluidstack': [],  # No dependencies needed for fluidstack
-    'cudo': ['cudo-compute>=0.1.10'],
-    'do': ['pydo>=0.3.0', 'azure-core>=1.24.0', 'azure-common'],
```
You can readd this to the new `sky/setup_files/dependencies.py`. Sorry for the merge conflict.

asaiacai commented:
@cblmemo smoke tests are passing on my end.  the catalog PR should fix the CI failures.

cblmemo commented:
> @cblmemo smoke tests are passing on my end. the catalog PR should fix the CI failures.

Thanks! QQ: Which catalog PR should I merge? Both skypilot-org/skypilot-catalog#83 and skypilot-org/skypilot-catalog#103?

Also, just want to make sure, does the CI passed in your side when merging the above mentioned catalog PRs?

asaiacai commented:
only skypilot-org/skypilot-catalog#103

pushed an update to the `test_optimizer_dryrun` since there are no `T4` or `32` memory instances on DO.
rest of the CI passes for me with this change

cblmemo commented on tests/test_optimizer_dryruns.py:
```diff
@@ -250,7 +257,12 @@ def test_instance_type_from_cpu_memory(monkeypatch, capfd):
-    _test_resources_launch(monkeypatch, memory=32)
+    enabled_clouds = list(clouds.CLOUD_REGISTRY.values())
+    # DO has no 32 mem instances
+    enabled_clouds.remove(clouds.CLOUD_REGISTRY['do'])
```
Will this cause an key error on user that has no do enabled?

asaiacai replied:
i removed this filter. resolving

cblmemo commented:
> only [skypilot-org/skypilot-catalog#103](https://github.com/skypilot-org/skypilot-catalog/pull/103)
> 
> pushed an update to the `test_optimizer_dryrun` since there are no `T4` or `32` memory instances on DO. rest of the CI passes for me with this change

Just curious, why `do` has no 32 memory instance? is it because all of them have higher memory?

Also, I think there is other cloud that has no T4 instance. Could you help check why the original test works for them but not `do`?

```bash
$ sky show-gpus T4 --cloud fluidstack
Resources 'T4' not found on Fluidstack. To show available accelerators, run: sky show-gpus --all
```

asaiacai commented:
> Just curious, why do has no 32 memory instance? is it because all of them have higher memory?

sorry for the back and forth here. The catalog was missing some instances with 32GB capability, which I just added to the catalog PR only skypilot-org/skypilot-catalog#103 I reverted `tests/test_optimizer_dryruns.py` back. so the last of the tests here are passing.

> Also, I think there is other cloud that has no T4 instance. Could you help check why the original test works for them but not do?
The filter against DO for the T4 case was removed, and the test case passes like `Fluidstack` now.

cblmemo commented:
Thanks @asaiacai ! Just merged the catalog PR. Re-running unittest now and after it passed it should be ready to go!
